src/bsoncxx/stdx/view_or_value.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace stdx {
+
+///
+/// Class representing a view-or-value variant type. Class View must be constructible
+/// from an instance of class Value, and class Value must be default constructible.
+///
+template <typename View, typename Value>
+class BSONCXX_API view_or_value {
+   public:
```

Comment:
also

`static_assert(std::is_constructible<View, Value>::value, "View type must be constructible from a value");`